pkg/util/active_groups.go

```diff
@@ -0,0 +1,156 @@
+	maxGroupsPerUser  int
+}
+
+func NewActiveGroups(maxGroupsPerUser int) *ActiveGroups {
+	return &ActiveGroups{
+		timestampsPerUser: map[string]map[string]*atomic.Int64{},
+		maxGroupsPerUser:  maxGroupsPerUser,
+	}
+}
+
+// UpdateGroupTimestampForUser function is only guaranteed to update to the
+// timestamp provided even if it is smaller than the existing value
+func (ag *ActiveGroups) UpdateGroupTimestampForUser(userID, group string, now time.Time) {
+	ts := now.UnixNano()
+	ag.mu.RLock()
+	if groupTs := ag.timestampsPerUser[userID][group]; groupTs != nil {
+		ag.mu.RUnlock()
+		groupTs.Store(ts)
+		return
+	}
+	ag.mu.RUnlock()
+
+	ag.mu.Lock()
+	defer ag.mu.Unlock()
+	groupTimestamps := ag.timestampsPerUser[userID]
+
+	if groupTimestamps == nil {
+		ag.timestampsPerUser[userID] = map[string]*atomic.Int64{group: atomic.NewInt64(ts)}
+		return
+	}
+
+	if groupTs := ag.timestampsPerUser[userID][group]; groupTs != nil {
+		groupTs.Store(ts)
+		return
+	}
+
+	ag.timestampsPerUser[userID][group] = atomic.NewInt64(ts)
+}
+
+func (ag *ActiveGroups) PurgeInactiveGroupsForUser(userID string, deadline int64) []string {
+	ag.mu.RLock()
+	totalGroups := len(ag.timestampsPerUser[userID])
+	if totalGroups == 0 {
+		ag.mu.RUnlock()
+		return nil
+	}
+
+	inactiveGroups := make([]string, 0, totalGroups)
+	groupTimestamps := ag.timestampsPerUser[userID]
+
+	for group, ts := range groupTimestamps {
+		if ts.Load() <= deadline {
+			inactiveGroups = append(inactiveGroups, group)
+		}
+	}
+	ag.mu.RUnlock()
+
+	if groupTimestamps == nil || len(inactiveGroups) == 0 {
+		return nil
+	}
+
+	// Cleanup inactive groups
+	deletedGroups := make([]string, 0, len(inactiveGroups))
+	ag.mu.Lock()
+	defer ag.mu.Unlock()
+
+	for _, inactiveGroup := range inactiveGroups {
+		groupTs := ag.timestampsPerUser[userID][inactiveGroup]
+		if groupTs != nil && groupTs.Load() <= deadline {
+			delete(ag.timestampsPerUser[userID], inactiveGroup)
+			deletedGroups = append(deletedGroups, inactiveGroup)
+		}
+	}
+
+	return deletedGroups
+}
+
+type ActiveGroupsCleanupService struct {
+	services.Service
+
+	activeGroups    *ActiveGroups
+	cleanupFuncs    []func(string, string) // Takes userID, group
+	inactiveTimeout time.Duration
+}
+
+type UserGroupMetricsCleaner interface {
+	RemoveGroupMetricsForUser(userID, group string)
+}
+
+func NewActiveGroupsCleanupService(cleanupInterval, inactiveTimeout time.Duration, maxGroupsPerUser int, cleanupFns ...func(string, string)) *ActiveGroupsCleanupService {
+	s := &ActiveGroupsCleanupService{
+		activeGroups:    NewActiveGroups(maxGroupsPerUser),
+		cleanupFuncs:    cleanupFns,
+		inactiveTimeout: inactiveTimeout,
+	}
+
+	s.Service = services.NewTimerService(cleanupInterval, nil, s.iteration, nil).WithName("active groups cleanup")
+	return s
+}
+
+func NewActiveGroupsCleanupWithDefaultValues(maxGroupsPerUser int, cleanupFns ...func(string, string)) *ActiveGroupsCleanupService {
```

Comment:
Who uses `cleanupFns`? If not used it, please remove it. I think it's just used in tests, but can call `Register()` instead, like "real" components do.